### PR TITLE
perf: compile ReAE stages and optimize compiled decoder layouts

### PR DESCRIPTION
**Pull request by gtong-nv**
# SwiftVR: compile ReAE stages and optimize compiled decoder layouts

## Summary

This PR adds an opt-in compiled ReAE path to the SwiftVR postprocessor and then
specializes the compiled decoder for the memory formats that TorchInductor and
cuDNN select efficiently on the measured GB300 system.

The default SwiftVR presets remain startup-friendly and numerically unchanged.
The existing `swiftvr-2x-compiled` preset opts into compiled transformer blocks
and the compiled ReAE decoder. Per-stream causal encoder/decoder state remains
outside compiled modules, so a resident pipeline can still serve isolated
streams without state sharing.

This description covers the two commits in this branch:

1. `acdaae94 perf: compile SwiftVR ReAE compute blocks`
2. `293a2932 Optimize compiled SwiftVR ReAE decoder layouts`

It deliberately excludes the currently uncommitted FP8 transformer experiment.

## Motivation

SwiftVR's ReAE encoder/decoder are a substantial fraction of end-to-end
restoration time. The existing eager implementation correctly preserves causal
stream state, but leaves stable convolution-heavy compute unfused. Earlier
profiling identified decoder work and layout conversions as higher-value targets
than input conversion, larger chunks, or attention-backend selection.

The goal is to improve steady-state 2x streaming throughput without changing
the default path or compromising causal output ordering and first-chunk trim
behavior.

### Commit 1: compile ReAE compute at stage boundaries

- Adds independent `compile_reae_encoder` and `compile_reae_decoder` controls
  to the SwiftVR pipeline, postprocessor configuration, and pipeline builder.
- Binds compiled encoder/decoder callables once when the resident model is
  created; they are not recreated per stream or per chunk.
- Keeps causal state explicit through the existing cache containers, rather
  than capturing it inside a compiled callable.
- Adds the opt-in `swiftvr-2x-compiled` preset, which compiles transformer
  blocks and the ReAE decoder. The encoder remains eager because compiling it
  alongside the decoder did not justify its extra preparation cost.
- Adds `benchmark_reae_compile.py`, a fresh-process benchmark harness that
  records warmup-excluded CUDA-event timings, p90, FPS, preparation time, peak
  allocation, restored tensors, and configuration metadata.
- Documents tested compile combinations, cold-start behavior, quality results,
  and rejected compile boundaries.

### Commit 2: layout-aware compiled decoder

- Adds `_SwiftVRCompiledDecoder`, a wrapper used only when
  `compile_reae_decoder=True`.
- Keeps ordinary decoder `Conv2d` regions and their inputs in
  `channels_last` format.
- Uses `channels_last_3d` for the temporal `Conv3d` boundary in
  `SwiftVRTemporalGrow`, then returns to contiguous layout before pixel-shuffle
  consumers.
- Preserves streaming semantics for every `MemBlock`: each block builds its
  one-frame left context from the state dictionary, updates that entry with the
  chunk's final frame, and runs the same residual convolution computation.
- Loads checkpoints before constructing the compiled wrapper, preserving the
  original checkpoint key mapping and eager/random-weight test path.
- Adds `benchmark_reae_layout.py` to compare contiguous, 2D channels-last, 3D
  channels-last, and combined layouts in separate processes/caches.
- Adds a CPU regression test that compares two streamed chunks against the
  baseline decoder and verifies temporal convolution weight layout.

## Performance results

Measurements use a single NVIDIA GB300, BF16, 16x16 shifted windows, zero DiT
overlap, eight 1280x704 RGB input frames per steady chunk, and 2560x1408
output. Values are evidence for this stack, not general guarantees.

### Compile sweep

| Candidate | Median / 8-frame chunk | FPS | Cold preparation |
| --- | ---: | ---: | ---: |
| Fully eager | 96.47 ms | 82.9 | 3.2 s |
| ReAE decoder compiled | 82.96 ms | 96.4 | 121.2 s |
| Transformer compiled | 94.69 ms | 84.5 | 30.0 s |
| Transformer + decoder compiled | **79.60 ms** | **100.5** | 142.0 s |

The selected preset improves steady-state latency by 17.5% versus the eager
baseline. It has a material cold-start cost, so it remains opt-in and should be
prewarmed in deployments that need first-output responsiveness.

### Decoder layout sweep

| Decoder layout | Eager median | Compiled median |
| --- | ---: | ---: |
| Contiguous | **32.03 ms** | 18.93 ms |
| Conv2d `channels_last` | 44.17 ms | 20.54 ms |
| Conv3d `channels_last_3d` | 32.87 ms | 10.91 ms |
| Combined Conv2d + Conv3d layout | 41.15 ms | **9.83 ms** |

The combined layout cuts isolated compiled decoder latency by 48.1% relative to
the compiled contiguous baseline. Explicit conversions make eager execution
slower, which is why this behavior is restricted to the compiled decoder.

### Full compiled-path confirmation

This comparison includes compiled transformer blocks and the ReAE decoder at
2560x1408; it is not an isolated decoder benchmark.

| Transformer + decoder configuration | Median / 8-frame chunk | FPS | Throughput gain |
| --- | ---: | ---: | ---: |
| Fully eager | 96.47 ms | 82.90 | baseline |
| Compiled transformer + contiguous compiled decoder | 80.56 ms | 99.30 | 19.79% |
| Compiled transformer + layout-aware compiled decoder | **71.89 ms** | **111.28** | **34.24%** |

The layout-aware path reduces full-path chunk latency by 25.48% (24.58 ms) and
raises throughput by 34.24% versus fully eager. It also reduces compiled-path
latency by 10.77% (8.67 ms) versus the contiguous compiled decoder. This
measurement uses BF16 and does not include the separate experimental FP8
transformer option.

## Correctness and quality

- Causal decoder state is still initialized by the wrapped decoder and remains
  keyed by each `MemBlock` identity.
- The layout regression test runs the baseline and layout-aware decoder over
  consecutive chunks and asserts identical streamed output on CPU.
- Direct compiled-decoder layout outputs were bit-identical in the isolated
  layout sweep.
- A full 24-frame pipeline comparison against the previous compiled decoder
  measured 0.000048 MAE, 0.000350 RMSE, and 69.13 dB PSNR, with no visible
  normal-scale difference.
- The registered compiled postprocessor smoke test produced exactly 24 finite,
  non-black 2560x1408 frames through buffered startup and flush.

## Validation

Executed for the layout follow-up:

```bash
uv run --no-sync pytest integrations_v2/swiftvr/tests/test_model.py -q
```

Result: `11 passed`.

The benchmark scripts are GPU/manual validation and intentionally use fresh
processes plus fresh `TORCHINDUCTOR_CACHE_DIR` locations for each candidate.

## Rollout notes

- No default preset behavior changes.
- `swiftvr-2x` remains the low-startup fallback.
- `swiftvr-2x-compiled` is appropriate for long-running or explicitly
  prewarmed streams, not latency-sensitive cold starts.
- The performance documentation includes exact benchmark methodology and
  rejected alternatives so future work can reproduce or revisit the decision.

### CODE REVIEW

File: integrations_v2/swiftvr/tests/test_model.py [hunk 1]
```diff
@@ -226,6 +236,79 @@ def test_reae_complete_group_adapter_preserves_stream_boundaries() -> None:
...
+        act_func=torch.nn.ReLU(inplace=True),
+    ).eval()
+    for index, block in enumerate(decoder.blocks):
+        if isinstance(block, TGrow):
+            decoder.blocks[index] = SwiftVRTemporalGrow(
+                int(block.conv.in_channels), block.stride
+            )
+
+    baseline = copy.deepcopy(decoder)
+    candidate = _SwiftVRCompiledDecoder(copy.deepcopy(decoder)).eval()
```
[the remark this answers is not in the document]
Comment (jmccaffrey-nv) [reply integrations_v2/swiftvr/tests/test_model.py:257]: accept

**Review by jmccaffrey-nv — APPROVED** [review]
